Create Hermes device that can use the xdma driver

iomartin opened the pull request:
Opening a new PR for the initial Hermes device to keep things cleaner. Original PR: #4

This PR introduces the Hermes device to QEMU.

This device has the 3 BARs defined on [eid-hermes][1] and it can do DMA operations using the [xdma][2] driver. It currently has one C2H and one H2C channel. This has been tested to work with the example on [3], including with large buffers that span multiple DMA descriptors.

In the future we will add eBPF capabilities to this device.

[1]: https://github.com/Eideticom/eid-hermes/blob/master/specs/eid-hermes-interface.md#overall-bar-layout
[2]: https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma
[3]: https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma#quick-example

iomartin committed:
Add libubpf to configure script

uBPF[1] is a userspace eBPF virtual machine, which includes an
assembler, disassembler, interpreter and JIT compiler for x86-64.

Because it is not part of the kernel, it can be used by userspace tools
that need to operate on programs in the eBPF ISA.

We need this to enable the hermes device model.

There is currently an open issue on uBPF[2] that requires the inclusion
of stdbool.h as well as ubpf.h. A PR has been opened[3], but has not yet
been merged into upstream, so we add that header to our configure script.

[1]: https://github.com/iovisor/ubpf
[2]: https://github.com/iovisor/ubpf/issues/38
[3]: https://github.com/iovisor/ubpf/pull/39

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Add minimal Hermes device

Only set Vendor ID and Device ID

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Add Hermes BAR 0

BAR 0 is used for hermes registers. Initially, it contains the device
capabilities. For the specification of BAR 0, please see [1].

BAR 0 is initialized with some arbitrary (but sensible) values.

[1]: https://github.com/Eideticom/eid-hermes/blob/master/specs/eid-hermes-interface.md

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Add Hermes BAR 4

BAR 4 is where a Hermes device stores eBPF programs and data for said
programs. Its layout consist of EHPSLOT program slots of size EHPSSZE
followed by EHDSLOT of size EHDSSZE. For more information, see [1].

BAR 4 should be programmed either via DMA commands sent by the host or
or via p2pdma (for data slots only).

[1]: https://github.com/Eideticom/eid-hermes/blob/master/specs/eid-hermes-interface.md

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Add Hermes BAR 2

BAR 2 contains the registers used by the XDMA driver. The layout of
this register is decribed on on Xilinx's manual PG195[1]. Note that that
document uses BAR1, but because on AWS F1 FPGAs it is mapped to BAR 2,
we use BAR 2 here as well.

This commits the registers, but does not perform DMAs yet. We only add
registers that are effectively used by the xdma driver[2].

We initialize registers to match AWS F1.

[1]: https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf
[2]: https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Add MSI-X init/uninit calls

We are going to use MSI-X to send interrupts to the xdma driver when a
DMA finishes. This has not been implemented yet.

For details on XDMA, see:

https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf
https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Implement DMA on Hermes

Using the xdma driver[1], we can now send DMA commands to Hermes. It has
been tested with this program[2] and works even if we increase the
buffer size (so it uses multiple descriptors, including when they cross
a 4K address boundary).

[1]: https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma
[2]: https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma#quick-example

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

iomartin committed:
Make the DMAs multi-threaded

A large DMA could freeze the guest while it is executed, so we need to
run them in other threads.

Create one thread for H2C and one for C2H.

Signed-off-by: Martin Oliveira <martin.oliveira@eideticom.com>

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -21,16 +21,87 @@
+    .ehpsoff = 0,
+    .ehpssze = (1 * MiB),
+    .ehdsoff = 16 * (1 * MiB), /* must be >= ehpslot * ehpssze */
+    .ehdssze = (1 * MiB)
```
Nits: the brackets are unnecessary, also it's good practice to add a trailing comma on the last item unless it's always going to be the last item (it's not here).

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -21,16 +21,87 @@
 
+#define HERMES_BAR0_SIZE          (32 * MiB)
+
+const struct hermes_bar0 {
```
Should be static.

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -57,11 +62,160 @@ const struct hermes_bar0 {
+    struct hermes_bar2_sgdma_reg c2h_sgdma;
+    struct hermes_bar2_sgdma_common_reg sgdma_common;
+    struct hermes_bar2_msix_pba_reg msix_pba;
+} bar2_init = {
```
Should be static... and I suspect const as well...

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -57,11 +62,160 @@ const struct hermes_bar0 {
+    struct hermes_bar2_msix_pba_reg msix_pba;
+} bar2_init = {
+    .h2c = {
+        .identifier = (0x1FC << 20) | (0x5),
```
I suspect we will need defines for a lot of these bits and magic numbers....

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -904,6 +914,53 @@ static const MemoryRegionOps hermes_bar2_ops = {
+    if (res < 0) {
+        trace_hermes_msix_init_fail(res);
+    } else {
+        if (!hermes_use_msix_vectors(hermes, HERMES_MSIX_VEC_NUM)) {
```
else if?

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -904,6 +914,53 @@ static const MemoryRegionOps hermes_bar2_ops = {
+                        &hermes->bar2_mem_reg, 2, HERMES_MSIX_PBA_OFFSET,
+                        0x0, &error_fatal);
+    if (res < 0) {
+        trace_hermes_msix_init_fail(res);
```
This doesn't seem right... trace events are not really good for communicating errors and realize() has a valid error path...

I'd suggest following the example in [vfio_msix_setup()](https://github.com/qemu/qemu/blob/4c41341af76cfc85b5a6c0f87de4838672ab9f89/hw/vfio/pci.c#L1504).

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+               | desc[num_desc - 1].nxt_addr_lo;
+    if (nxt_addr) {
+        nxt_adj = (desc[num_desc - 1].ctrl >> 8) & 0x3F;
+        num_desc += __do_dma(hermes, nxt_addr, nxt_adj + 1, h2c);
```
__do_dma() can return -1.... so we should probably be careful about error handling here...

lsgunth replied:
Also, recursion is probably a really bad idea here... as it usually is...

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+    num_desc = 1 + (sgdma_reg->desc_num_adj & 0x3F);
+    desc_addr = ((uint64_t) sgdma_reg->desc_high_addr) << 32 |
+                sgdma_reg->desc_low_addr;
+    num_desc = __do_dma(hermes, desc_addr, num_desc, h2c);
```
__do_dma can return -1... so we need to be careful about the returned errors here

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
 {
 }
 
+static int __do_dma(struct HermesState *hermes, hwaddr desc_addr,
```
I don't think the struct keyword here is correct given the use of the typedef in other places...

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+    /* DMA all descriptors in this block */
+    for (int i = 0; i < num_desc; i++) {
+        src_addr = ((uint64_t) desc[i].src_addr_hi << 32) | desc[i].src_addr_lo;
+        dst_addr = ((uint64_t) desc[i].dst_addr_hi << 32) | desc[i].dst_addr_lo;
```
Is there a reason these can't just be 64bit types?

iomartin replied:
Nope, fixed

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -57,11 +62,160 @@ const struct hermes_bar0 {
+struct hermes_bar2_sgdma_reg {
+    uint32_t identifier;     /* 0x00 */
+    uint32_t desc_low_addr;  /* 0x80 */
+    uint32_t desc_high_addr; /* 0x84 */
```
Given that these are internal to the model, I see no reason why you can't store the address as a uint64... or even hwaddr immediately instead of storing an upper and a lower and combining them at use.

iomartin replied:
Makes sense, fixed

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+    unsigned irq_vector;
+    unsigned num_desc;
+
+    if (h2c) {
```
Still don't like these h2c bools... Better to restructure your data a bit to make this cleaner...

This very rough, certainly won't compile mockup should hopefully demonstrate what I'm thinking:

```diff
diff --git a/hw/misc/hermes.c b/hw/misc/hermes.c
index 6c65c458c0a9..aa125b6c2b57 100644
--- a/hw/misc/hermes.c
+++ b/hw/misc/hermes.c
@@ -83,7 +83,7 @@ typedef struct HermesState HermesState;
  * https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf
  * (note that in this document it uses BAR1, but here and on AWS we use BAR2)
  */
-struct hermes_bar2_engine_reg {
+struct hermes_bar2_channel_reg {
     uint32_t identifier;         /* 0x00 */
     uint32_t control;            /* 0x04, 0x08 and 0x0C */
     uint32_t status;             /* 0x40 and 0x44 */
@@ -153,13 +153,17 @@ struct hermes_bar2_msix_pba_reg {
     uint32_t pba;                   /* 0xFE0 */
 };
 
+struct hermes_bar2_dir_reg {
+    struct hermes_bar2_engine_reg channel;
+    struct hermes_bar2_sgdma_reg sgdma;
+    void (*perform_dma)(HermesState hermes, hwaddr src_addr, dst_addr);
+    int irq_vector;
+};
+
 struct hermes_bar2 {
-    struct hermes_bar2_engine_reg h2c;
-    struct hermes_bar2_engine_reg c2h;
+    struct hermes_bar2_dir_reg h2c, c2h;
     struct hermes_bar2_irq_reg irq;
     struct hermes_bar2_cfg_reg cfg;
-    struct hermes_bar2_sgdma_reg h2c_sgdma;
-    struct hermes_bar2_sgdma_reg c2h_sgdma;
     struct hermes_bar2_sgdma_common_reg sgdma_common;
     struct hermes_bar2_msix_pba_reg msix_pba;
 } bar2_init = {
@@ -274,110 +278,109 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
 {
 }
 
-static int __do_dma(struct HermesState *hermes, hwaddr desc_addr,
-                       uint8_t num_desc, bool h2c)
+static void hermes_h2c_dma(HermesState hermes, hwaddr src_addr, dst_addr)
+{
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    trace_hermes_dma("H2C", src_addr, dst_addr);
+    pci_dma_read(&hermes->pdev, src_addr, bar4_base + dst_addr,
+                 desc[i].len);
+
+}
+
+static void hermes_c2h_dma(HermesState hermes, hwaddr src_addr, dst_addr)
+{
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    trace_hermes_dma("C2H", src_addr, dst_addr);
+    pci_dma_write(&hermes->pdev, dst_addr, bar4_base + src_addr,
+                  desc[i].len);
+}
+
+static int hermes_execute_descs(struct HermesState *hermes,
+		struct hermes_bar2_dir_reg *dir, hwaddr *desc_addr, uint8_t *num_desc)
 {
     struct hermes_dma_desc *desc;
     hwaddr src_addr, dst_addr, nxt_addr;
     uint8_t nxt_adj;
-    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
 
-    trace_hermes_dma_num_adj(num_desc - 1);
-    desc = g_new(struct hermes_dma_desc, num_desc);
+    trace_hermes_dma_num_adj(*num_desc - 1);
+    desc = g_new(struct hermes_dma_desc, *num_desc);
     if (!desc) {
         fprintf(stderr, "[Hermes] Failed to alloc memory for DMA descriptor\n");
         return -1;
     }
 
     /* Read DMA descriptors */
-    pci_dma_read(&hermes->pdev, desc_addr, desc, num_desc * sizeof(*desc));
-    for (int i = 0; i < num_desc; i++) {
+    pci_dma_read(&hermes->pdev, *desc_addr, desc, *num_desc * sizeof(*desc));
+    for (int i = 0; i < *num_desc; i++) {
         trace_hermes_dma_desc(desc[i].ctrl, desc[i].len, desc[i].src_addr_lo,
                               desc[i].src_addr_hi, desc[i].dst_addr_lo,
                               desc[i].dst_addr_hi, desc[i].nxt_addr_lo,
                               desc[i].nxt_addr_hi);
-    }
 
-    /* DMA all descriptors in this block */
-    for (int i = 0; i < num_desc; i++) {
         src_addr = ((uint64_t) desc[i].src_addr_hi << 32) | desc[i].src_addr_lo;
         dst_addr = ((uint64_t) desc[i].dst_addr_hi << 32) | desc[i].dst_addr_lo;
 
-        if (h2c) {
-            trace_hermes_dma("H2C", src_addr, dst_addr);
-            pci_dma_read(&hermes->pdev, src_addr, bar4_base + dst_addr,
-                         desc[i].len);
-        } else {
-            trace_hermes_dma("C2H", src_addr, dst_addr);
-            pci_dma_write(&hermes->pdev, dst_addr, bar4_base + src_addr,
-                          desc[i].len);
-        }
+        perform_dma(hermes, src_addr, dst_addr);
     }
 
-    nxt_addr = ((uint64_t) desc[num_desc - 1].nxt_addr_hi << 32)
-               | desc[num_desc - 1].nxt_addr_lo;
-    if (nxt_addr) {
-        nxt_adj = (desc[num_desc - 1].ctrl >> 8) & 0x3F;
-        num_desc += __do_dma(hermes, nxt_addr, nxt_adj + 1, h2c);
-    }
+    *desc_addr = ((uint64_t) desc[i - 1].nxt_addr_hi << 32)
+                 | desc[i - 1].nxt_addr_lo;
+    *num_desc = (desc[i - 1].ctrl >> 8) & 0x3F
 
     g_free(desc);
 
-    return num_desc;
+    return i;
 }
 
-static void do_dma(HermesState *hermes, bool h2c)
+static void hermes_do_dma(HermesState *hermes, struct hermes_bar2_dir_reg *dir)
 {
-    struct hermes_bar2_engine_reg *engine_reg;
-    struct hermes_bar2_sgdma_reg *sgdma_reg;
     struct hermes_bar2 *bar2 = &hermes->bar2;
     struct hermes_bar2_irq_reg *irq = &bar2->irq;
     hwaddr desc_addr;
     unsigned irq_vector;
     unsigned num_desc;
-
-    if (h2c) {
-        sgdma_reg = &bar2->h2c_sgdma;
-        engine_reg = &bar2->h2c;
-        irq_vector = 0;
-    } else {
-        sgdma_reg = &bar2->c2h_sgdma;
-        engine_reg = &bar2->c2h;
-        irq_vector = 1;
-    }
+    int ret;
 
     /* Reset number of completed descriptors */
-    engine_reg->cmp_desc_count = 0;
+    dir->channel.cmp_desc_count = 0;
 
     /* Set engine as busy */
-    atomic_or(&engine_reg->status, 0x1);
+    atomic_or(&dir->channel.status, 0x1);
 
     /*
      * There is always at least one descriptor, plus the adjacent ones (which
      * could be 0). Only bits 5:0 of the register are defined
      */
-    num_desc = 1 + (sgdma_reg->desc_num_adj & 0x3F);
-    desc_addr = ((uint64_t) sgdma_reg->desc_high_addr) << 32 |
-                sgdma_reg->desc_low_addr;
-    num_desc = __do_dma(hermes, desc_addr, num_desc, h2c);
+    num_desc = 1 + (dir->sgdma.desc_num_adj & 0x3F);
+    desc_addr = ((uint64_t) dir->sgdma->desc_high_addr) << 32 |
+        		dir->sgdma->desc_low_addr;
+    engine_reg->cmp_desc_count = 0;
+
+    while (1) {
+        ret = hermes_execute_descs(hermes, dir, &desc_addr, &num_desc);
+        if (ret < 0 || !desc_addr)
+            break;
 
-    /* Set number of completed descriptors */
-    engine_reg->cmp_desc_count = num_desc;
+        /* Set number of completed descriptors */
+        dir->channel.cmp_desc_count += ret;
+    }
 
     /* Set engine as not busy */
-    atomic_and(&engine_reg->status, ~1);
+    atomic_and(&dir->channel.status, ~1);
 
-    if (irq->chnl_inter_enable_mask & (irq_vector + 1)) {
+    if (irq->chnl_inter_enable_mask & (dir->irq_vector + 1)) {
         /* Set interrupt source */
         irq->chnl_inter_request =  irq->chnl_inter_enable_mask &
-                                   (irq_vector + 1);
+                                   (dir->irq_vector + 1);
 
         /*
          * Send interrupt. Since we currently have only one channel for H2C and
          * one for C2H, we have that IRQ 0 is H2C and IRQ 1 is C2H
          */
         trace_hermes_msix_notify(irq_vector);
-        msix_notify(PCI_DEVICE(hermes), irq_vector);
+        msix_notify(PCI_DEVICE(hermes), dir->irq_vector);
     }
 }


```

iomartin replied:
Thanks for the suggestion and code. Fixing it.

lsgunth reviewed (commented):
On general point I'd make for BAR2, is to maybe move away from the structures and data model being so tied with the address space register definitions... Seeing they are not actually tied to them at all.

There seem to be a lot of registers that aren't used (and perhaps are read only) that it doesn't seem like we need to store state for and they could just be hardcoded in the switch statements. I'd to focus the code on what's needed for the model, not implementing every unused register.

Often times it's easiest to understand code by looking at it's data structures. However, it's difficult here seeing there's a lot of data that doesn't seem to do anything.... Though perhaps this will be expanded on?

Also, I'd like to say this XDMA design seems super unusual and not well optimized. Seems to simply be naively designed to take a single, inefficiently specified SGL with no means to queue work or anything like that. Among other awkwardness especially given what their kernel driver does. Seems like it wasn't clear on the use case and just ended up being an odd set of choices as far as DMA engines go.

iomartin commented:
Force-pushed with many changes. I have updated the DMA code to do something similar to what @lsgunth suggested. I also traced all BAR2 accesses that the XDMA driver did and removed all registers that weren't being used (and added unimplemented warnings for them).

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -248,6 +279,88 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
+
+    /* Read DMA descriptors */
+    pci_dma_read(&hermes->pdev, *desc_addr, desc, *num_desc * sizeof(*desc));
+    for (i = 0; i < *num_desc; i++) {
```
What's the point of putting all the traces in their own loop?

iomartin replied:
Ops, missed that

lsgunth commented on hw/misc/hermes.c:
```diff
@@ -960,12 +995,35 @@ static void pci_hermes_realize(PCIDevice *pdev, Error **errp)
     hermes_init_msix(hermes, errp);
+
+    hermes->stopping = false;
+    qemu_mutex_init(&hermes->bar2.h2c.dma_mutex);
```
Might be worth having a function to init and destroy the dir_regs so we don't have to double up on each init and have long hermes->bar2.... prefixes on every variable.

lsgunth reviewed (approved):
Looks a lot cleaner. Couple small nits.

sbates130272 commented:
Looks good @iomartin. Please clean up the last couple of NITs from @lsgunth and then feel free to merge!

sbates130272 reviewed (approved):
Looks good!
